### PR TITLE
Simplified comparison update

### PR DESCRIPTION
**Pull request by RichardDalton**
With just the Func and the check for IEquatable.

Replacing [previous pull request](https://github.com/Haacked/Scientist.net/pull/2)

### CODE REVIEW

File: UnitTests/ComplexResult.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+
+public class ComplexResult : IEquatable<ComplexResult>
```
Comment (Haacked) [inline UnitTests/ComplexResult.cs]: Ha! I kept wondering why this class is here and then I realized, this is for the unit tests! Silly me.

File: UnitTests/ScientistTests.cs [hunk 2]
```diff
@@ -121,5 +122,103 @@ public void AnExceptionReportsDuration()
...
+
+        [Fact]
+        public void RunsBothBranchesOfTheExperimentWithResultComparisonSetAndReportsSuccess()
+        {
+            bool candidateRan = false;
+            bool controlRan = false;
+
+            // We introduce side effects for testing. Don't do this in real life please.
+            // Do we do a deep comparison?
+            Func<ComplexResult> control = () => { controlRan = true; return new ComplexResult {Count = 10, Name = "Tester"}; };
```
Comment (Haacked) [inline UnitTests/ScientistTests.cs]: Now that we are using NSubstitute, want to take a crack at converting this test to use that? Check out the other tests for examples.

File: UnitTests/ComplexResult.cs (hunk 1 above)
Comment (RichardDalton) [inline UnitTests/ComplexResult.cs]: Maybe I should have named it better. Damn those really hard problems!

File: src/Scientist/Internals/Experiment.cs [hunk 3]
```diff
@@ -30,6 +31,14 @@ public ExperimentInstance(string name, Func<Task<T>> control, Func<Task<T>> cand
             _candidate = candidate;
         }
 
+        public ExperimentInstance(string name, Func<Task<T>> control, Func<Task<T>> candidate, Func<T, T, bool> resultComparison)
```
Comment (Haacked) [inline src/Scientist/Internals/Experiment.cs]: Seems like `resultComparer` would be a better name. To me, "comparison" sounds like the result of comparing. Thoughts?

File: src/Scientist/Internals/Experiment.cs [hunk 4]
```diff
@@ -66,13 +72,30 @@ public ExperimentInstance(string name, Func<Task<T>> control, Func<Task<T>> cand
             return controlResult.Result;
         }
 
+        private bool CompareResults(ExperimentResult controlResult, ExperimentResult candidateResult)
+        {
+            if (_resultComparison != null)
+            {
+                return _resultComparison(controlResult.Result, candidateResult.Result);
```
Comment (Haacked) [inline src/Scientist/Internals/Experiment.cs]: As I think about this more, I worry taht people supplying a comparer function will forget to handle null checks. In fact, why should we make them?

If exactly one of the results is null and the other is not, then we know to return false.
If both results are null, we return true.

If both results are not null, we then continue and use the result comparers etc.

That way, developers who pass in a `resultComparison` don't need to worry about checking for null. The tricky part is that `T` is not constrained to be a reference type. So we'd probably need to cast to `object` and then do the null checks. That leads to boxing in the case of value types, which sucks, but I think it makes the API easier to use. Thoughts?

File: src/Scientist/Internals/Experiment.cs [hunk 5]
```diff
@@ -14,6 +14,7 @@ internal class ExperimentInstance<T>
 
         readonly Func<Task<T>> _control;
         readonly Func<Task<T>> _candidate;
+        private readonly Func<T, T, bool> _resultComparison;
```
Comment (Haacked) [inline src/Scientist/Internals/Experiment.cs]: Mind removing `private` to adhere to the current convention?

**Comment by Haacked** [comment]
:pineapple: Done with my review. Some comments for you.

File: UnitTests/ComplexResult.cs (hunk 1 above)
Comment (Haacked) [inline UnitTests/ComplexResult.cs]: Nah, it's good. I just needed to look at where it lives.

**Comment by RichardDalton** [comment]
Cheers, will make those changes now.

**Comment by RichardDalton** [comment]
Changes done. In the CompareResults method I have moved the null checks to the top so any consumers do not need to worry about null checking the objects that are passed to their comparison code.

**Comment by Haacked** [comment]
I'll fix up these conflicts as I want to make some slight changes.

**Pull request merged**